### PR TITLE
fix(api): Handle smoothie update better

### PR DESCRIPTION
**Pull request by sfoster1**
The smoothie update detection machinery assumes that when the server starts, if
the smoothie does not respond it must be in programming mode. However, this can
also be true if the api server is booting for the first time and expects a
smoothie version that isn't present, like the recent change. In that case, the
update will fail because the smoothie is never set to programming mode. This
commit adds a couple retries for smoothei update, and forces programming mode
for attempts past the first.

In addition, failed firmware updates currently could in theory make the api
server stop because of an uncaught exception. Instead, force the smoothie to be virtual.

## Testing
- Use this to update some robots still on edge-616 smoothie firmware and make sure it works
- See if you can force the smoothie into programming mode when booting and make sure that case still works too

### CODE REVIEW

File: api/src/opentrons/legacy_api/robot/robot.py [hunk 1]
```diff
@@ -1052,6 +1052,7 @@ def update_config(self, **kwargs):
         msg = await self._driver.update_firmware(
             filename, checked_loop, explicit_modeset)
         self.fw_version = self._driver.get_fw_version()
+        self.connect()
```
Comment (Laura-Danielle) [inline api/src/opentrons/legacy_api/robot/robot.py]: what's the purpose of connecting to the robot after firmware update?

Comment (sfoster1) [reply api/src/opentrons/legacy_api/robot/robot.py]: Running the initial pipette config, I guess it's not strictly necessary

File: api/src/opentrons/drivers/smoothie_drivers/driver_3_0.py [hunk 2]
```diff
@@ -1564,14 +1566,22 @@ def _smoothie_hard_halt(self):
...
         proc = await asyncio.create_subprocess_shell(
             update_cmd, **kwargs)
-        rd: bytes = await proc.stdout.read()  # type: ignore
-        res = rd.decode().strip()
-        await proc.communicate()
+        out_b, err_b = await proc.communicate()
+        if proc.returncode != 0:
+            log.error(
+                f"Smoothie update failed: {proc.returncode} {out_b} {err_b}")
+            raise RuntimeError(
```
Comment (mcous) [inline api/src/opentrons/drivers/smoothie_drivers/driver_3_0.py]: Will this `raise` cause any problems outside of the boot update? Does this method get called by the firmware update endpoint?

Comment (sfoster1) [reply api/src/opentrons/drivers/smoothie_drivers/driver_3_0.py]: Sadly this method is not called by the firmware update endpoint; instead that endpoint separately invokes the programmer executable (https://github.com/Opentrons/opentrons/blob/ebc41a472e31539fe2ebf31370e6dfdf4b786439/update-server/otupdate/balena/install.py#L95)

**Pull request merged**